### PR TITLE
chore: 🆙 upgrade typescript-eslint

### PR DESCRIPTION
**Pull request by afc163**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [x] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
Need help to fix new eslint errors.

**Comment by abenhamdine** [comment]
I would recommend not upgrading to typescript-eslint 4 until these annoying false positives are fixed, see https://github.com/typescript-eslint/typescript-eslint/milestone/7

**Comment by Kermit-Xuan** [comment]
image

Obviously this isn't problems of code.

I think we can upgrade them when they are stable.

**Comment by abenhamdine** [comment]
For information, https://github.com/typescript-eslint/typescript-eslint/releases/tag/v4.1.0 has been released and fixes 90% of v4 false positives, but I still see a few remaining false positives in our codebase.

**Comment by yoyo837** [comment]
Need to continue to wait for version 4.1.1 or higher...

**Comment by abenhamdine** [comment]
Actually, community feedbacks show there more and more false positives detected, 4.X is definitely not currently usable.

**Comment by hengkx** [comment]
#2540

File: components/modal/confirm.tsx [hunk 1]
```diff
@@ -33,7 +33,7 @@ export interface ModalStaticFunctions {
 export default function confirm(config: ModalFuncProps) {
   const div = document.createElement('div');
   document.body.appendChild(div);
-  // eslint-disable-next-line no-use-before-define
+  // eslint-disable-next-line @typescript-eslint/no-use-before-define
   let currentConfig = { ...config, close, visible: true } as any;
```
Comment (hengkx) [inline components/modal/confirm.tsx:37]: 我也想知道 close 是从什么地方 来的 没看到

Comment (Kermit-Xuan) [reply components/modal/confirm.tsx:37]: window.close?

**Comment by hengkx** [comment]
lint 终于 过了🥳

File: components/transfer/__tests__/index.test.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/* eslint no-use-before-define: "off" */
+/* eslint @typescript-eslint/no-use-before-define: "off" */
```
Comment (afc163) [inline components/transfer/__tests__/index.test.js:1]: 这个还需要么

Comment (hengkx) [reply components/transfer/__tests__/index.test.js:1]: 需要

File: components/modal/confirm.tsx (hunk 1 above)
Comment (yoyo837) [reply components/modal/confirm.tsx:37]: > window.close?

不是，就是一个close的props

Comment (afc163) [reply components/modal/confirm.tsx:37]: 可以提到前面来：

https://github.com/ant-design/ant-design/blob/9d676eb7c0e7236e2517d576ce3342a12b0dcea1/components/modal/confirm.tsx#L78-L85

File: .eslintrc.js [hunk 3]
```diff
@@ -121,6 +121,14 @@ module.exports = {
     'unicorn/prefer-trim-start-end': 2,
     'unicorn/expiring-todo-comments': 2,
     'unicorn/no-abusive-eslint-disable': 2,
+
+    // https://github.com/typescript-eslint/typescript-eslint/issues/2540#issuecomment-692866111
+    'no-use-before-define': 0,
```
Comment (yoyo837) [inline .eslintrc.js:126]: 为什么他们不通过覆盖config的方式来关闭，要用户自己配置呢？

File: components/modal/confirm.tsx (hunk 1 above)
Comment (hengkx) [reply components/modal/confirm.tsx:37]: > > window.close?
> 
> 不是，就是一个close的props

根本 没有 定义。。。

File: .eslintrc.js (hunk 3 above)
Comment (hengkx) [reply .eslintrc.js:126]: 我也不知道  也没 找到 其它 好的 方案了 ，只能 这样先 干掉，看了下 ts 的确 是正确的 引用

File: components/modal/confirm.tsx (hunk 1 above)
Comment (afc163) [reply components/modal/confirm.tsx:37]: 看我的回复。

Comment (hengkx) [reply components/modal/confirm.tsx:37]: 提不了吧 互相 引用了

**Pull request merged**